[docs] Add section about overriding slots to base concepts (#10421)
mui/mui-x: docs/data/date-pickers/base-concepts/base-concepts.md

// Example of a test using the helper
expect(cleanText(input.value)).to.equal('04-17-2022');
```

> ## Overriding `slots` and `slotProps`

Reviewer: Avoid code blocks in headers:

```suggestion
## Overriding slots and slot props
```

> Date and Time Pickers are complex components built using many subcomponents. These internal elements can be overridden using the `slots` prop, allowing for easy customization. You can also pass additional props to the available `slots`, using the `slotProps` prop.

Reviewer: Passive —> active voice

```suggestion
Date and Time Pickers are complex components built using many subcomponents. You can override these internal elements using the `slots` prop, allowing for easy customization. You can also pass additional props to the available `slots` using the `slotProps` prop.
```

> The list of available `slots` for each component can be found on the component's [API Reference](/x/api/date-pickers/date-picker/#slots) page.

Reviewer: Active voice

```suggestion
You can find the list of available `slots` for each component in its respective [API reference](/x/api/date-pickers/date-picker/#slots) doc.
```

> Some parts of the pickers' UI are built on several nested slots. For instance, the adornment of the `TextField` on `DatePicker` contains three slots (`inputAdornment`, `openPickerButton`, and `openPickerIcon`) that you can use depending on what you are trying to customize.

Reviewer: - capitalize Pickers since we're referring to the product, not the general concept

```suggestion
Some parts of the Pickers' UI are built on several nested slots. For instance, the adornment of the `TextField` on `DatePicker` contains three slots (`inputAdornment`, `openPickerButton`, and `openPickerIcon`) that you can use depending on what you are trying to customize.
```

{{"demo": "CustomSlots.js"}}

:::info

> Visit the [dedicated documentation page](/x/react-date-pickers/custom-components/) for more information about overriding `slots`.

Reviewer: informative link

```suggestion
Learn more about overriding slots in the [Custom components doc](/x/react-date-pickers/custom-components/).
```

// Example of a test using the helper
expect(cleanText(input.value)).to.equal('04-17-2022');
```

> ## Overriding `slots` and `slotProps`

Reviewer 2: Should be fixed by https://github.com/mui/material-ui/pull/39165

**Slots** are the internal building blocks of the components. These internal elements can be overriden using the `slots` prop, allowing for easy customization of Date and Time Pickers. You can also pass additional props to the available `slots`, using the `slotProps` prop.

The list of available `slots` for each component can be found in the [API References](/x/api/date-pickers/).

You can easily cutomize simple slots as the `openPickerIcon` just by replacing the slot like this:

```jsx
<DatePicker slots={{ openPickerIcon: EditCalendarRoundedIcon }} />
```

> The subcomponents of the pickers are often built on more nested structures, and you can go as deep into the customization of the nested elements as you want. For instance, you can customize the `IconButton` used as the `openPickerButton`, as well as the `openPickerIcon` inside of it. You can do the same for the `field` and the `textField`.

Reviewer 3: IMHO, this section is a bit misleading. You could get the impression that we expose every single component as a slot, even though it's far from the truth. 🤔 
I don't have a much better suggestion for this section, but maybe someone else could pitch in? 🤔

Author: Hmmm...I see what you mean. Maybe including something like 'some of which can be overridden'? To make it clear that not everything can be replaced 🤔 I don't really have a better idea
cc. @flaviendelangle @joserodolfofreitas

Reviewer 4: Maybe we can take it from another angle

> Some parts of the UI contains several nested slots. For instance, the adornment of the `TextField` contains 3 slots ([`inputAdornment`, `openPickerButton`, `openPickerIcon`](/x/react-date-pickers/custom-opening-button/) that you can use depending on the React element you are trying to custimize.

Author: WDYT about the current change?
